reducer/src/main/java/com/graphicsfuzz/reducer/reductionopportunities/RemoveRedundantUniformMetadataReductionOpportunities.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright 2019 The GraphicsFuzz Project Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.graphicsfuzz.reducer.reductionopportunities;
+
+import com.graphicsfuzz.common.ast.TranslationUnit;
+import com.graphicsfuzz.common.ast.decl.VariableDeclInfo;
+import com.graphicsfuzz.common.ast.decl.VariablesDeclaration;
+import com.graphicsfuzz.common.ast.expr.VariableIdentifierExpr;
+import com.graphicsfuzz.common.transformreduce.ShaderJob;
+import com.graphicsfuzz.common.util.ListConcat;
+import com.graphicsfuzz.common.util.PipelineInfo;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+public class RemoveRedundantUniformMetadataReductionOpportunities
+    extends ReductionOpportunitiesBase<RemoveRedundantUniformMetadataReductionOpportunity> {
```

Comment:
I am unsure whether this class should extend ReductionOpportunitiesBase after all, since it is fundamentally a bit different: it may need to consider multiple shaders in order to know what the opportunities are (since a uniform might be used in a vertex shader but not fragment shader, for example).

I think it might be preferable to have a class that implements IReductionOpportunity directly.

Reply:
Fixed